autosklearn/data/feature_validator.py

```diff
@@ -0,0 +1,399 @@
+                    elif cmp1 == 'numerical' and cmp2 == 'categorical':
+                        return 1
+                    else:
+                        raise ValueError((cmp1, cmp2))
+                self.feature_types = sorted(
+                    self.feature_types,
+                    key=functools.cmp_to_key(comparator)
+                )
+        return self
+
+    def transform(
+        self,
+        X: typing.Union[np.ndarray, pd.DataFrame],
+    ) -> np.ndarray:
+        """
+        Validates and fit a categorical encoder (if needed) to the features.
+        The supported data types are List, numpy arrays and pandas DataFrames.
+
+        Parameters
+        ----------
+            X_train: typing.Union[np.ndarray, pd.DataFrame]
+                A set of features, whose categorical features are going to be
+                transformed
+
+        Return
+        ------
+            np.ndarray:
+                The transformed array
+        """
+        if not self._is_fitted:
+            raise ValueError("Cannot call transform on a validator that is not fitted")
+
+        # Pandas related transformations
+        if hasattr(X, "iloc") and self.encoder is not None:
+            if np.any(pd.isnull(X)):
+                # After above check it means that if there is a NaN
+                # the whole column must be NaN
+                # Make sure it is numerical and let the pipeline handle it
+                for column in X.columns:
+                    if X[column].isna().all():
+                        X[column] = pd.to_numeric(X[column])
+            try:
+                X = self.encoder.transform(X)
+            except ValueError as e:
+                if 'Found unknown categories' in e.args[0]:
+                    # Make the message more informative
+                    raise ValueError(
+                        "During fit, the input features contained categorical values in columns"
+                        "{}, with categories {} which were encoded by Auto-sklearn automatically."
+                        "Nevertheless, a new input contained new categories not seen during "
+                        "training = {}. The OrdinalEncoder used by Auto-sklearn cannot handle "
+                        "this yet (due to a limitation on scikit-learn being addressed via:"
+                        " https://github.com/scikit-learn/scikit-learn/issues/17123)"
+                        "".format(
+                            self.enc_columns,
+                            self.encoder.transformers_[0][1].categories_,
+                            e.args[0],
+                        )
+                    )
+                else:
+                    raise e
+
+        # Sparse related transformations
+        if scipy.sparse.issparse(X):
+            X.sort_indices()
+
+        return sklearn.utils.check_array(
+            X,
+            force_all_finite=False,
+            accept_sparse='csr'
+        )
+
+    def check_data(
```

Comment:
Is there a reason this is not part of `_fit()`? It is a bit confusing to me that this is in a separate function.

Reply:
I have moved this to the fit() function